### ISSUE TITLE
Keep selection information in the response

### ISSUE THREAD

**Issue by AtnNn** [rethinkdb/rethinkdb#523]
I would add optional database name, table name and primary key fields to the `Response` protobuf message. These fields would be filled in whenever the query is a Selection.

This would allow drivers to associate each returned object with a document in the database. My haskell driver currently does this by keeping track of what table is being queried, but it make things very complicated and fails with code like `r.branch(foo, r.table('a'), r.table('b'))`. It also assumes that the primary key is `id`.

Here is a simple example from the [google datastore tutorial](https://developers.google.com/appengine/docs/python/datastore/overview) that illustrates their API. In this example, `e` is the object returned by the query, but it also remembers the document it refers to.

```
employees_trained = db.GqlQuery("SELECT * FROM Employee WHERE email IN :1",
                                training_registration_list)
for e in employees_trained:
  e.new_hire_training_completed = True
  db.put(e)
```

**Comment by srh** [comment]
This would be some shadow-channel into ReQL that isn't accessible by the main query language, only the client.  Instead it should be accessible from ReQL, and _not_ returned to the client.  We should max out our performance, after all...  There's some related discussion in #6267 about attaching "is from a table" metadata to documents too, for use in `.sindex_val()` or such.

**Comment by marshall007** [comment]
@srh agreed, implementing this via the context object discussed in #6267 is the way to go on this one.
